prover/circuit_utils.go

```diff
@@ -33,31 +34,180 @@ func (e *bitPatternLengthError) Error() string {
+	root = extractor.Call(api, VerifyProof{Proof: proof, Path: currentPath})
+
+	return root
+}
+
+type InsertionProof struct {
+	StartIndex frontend.Variable
+	PreRoot    frontend.Variable
+	IdComms    []frontend.Variable
+
+	MerkleProofs [][]frontend.Variable
+
+	BatchSize int
+	Depth     int
+}
+
+func (gadget InsertionProof) DefineGadget(api abstractor.API) interface{} {
+	prevRoot := gadget.PreRoot
+
+	// Individual insertions.
+	for i := 0; i < gadget.BatchSize; i += 1 {
+		currentIndex := api.Add(gadget.StartIndex, i)
+		prevRoot = extractor.Call(api, InsertionRound{
+			Index: currentIndex,
+			Item: gadget.IdComms[i],
+			PrevRoot: prevRoot,
+			Proof: gadget.MerkleProofs[i],
+			Depth: gadget.Depth,
+		})
+	}
+
+	return prevRoot
+}
+
+type DeletionRound struct {
+	Root          frontend.Variable
+	Index         frontend.Variable
+	Item          frontend.Variable
+	MerkleProofs  []frontend.Variable
+
+	Depth         int
+}
+
+func (gadget DeletionRound) DefineGadget(api abstractor.API) interface{} {
+	// We verify that the Leaf belongs to the Merkle Tree by verifying that the computed root
+	// matches gadget.Root. Then, we return the root computed with Leaf being empty.
+	currentPath := api.ToBinary(gadget.Index, gadget.Depth+1)
+	// Treating indices with the one-too-high bit set as a skip flag. This allows
+	// us to pad batches with meaningless ops to commit something even if there
+	// isn't enough deletions happening to fill a batch.
+	skipFlag := currentPath[gadget.Depth]
+	currentPath = currentPath[:gadget.Depth]
+
+	// Verify proof for Item.
+	rootPreDeletion := extractor.Call(api, VerifyProof{append([]frontend.Variable{gadget.Item}, gadget.MerkleProofs[:]...), currentPath})
+
+	// Verify proof for empty leaf.
+	rootPostDeletion := extractor.Call(api, VerifyProof{append([]frontend.Variable{emptyLeaf}, gadget.MerkleProofs[:]...), currentPath})
+
+	preRootCorrect := api.IsZero(api.Sub(rootPreDeletion, gadget.Root))
+	preRootCorrectOrSkip := api.Or(preRootCorrect, skipFlag)
+	api.AssertIsEqual(preRootCorrectOrSkip, 1)
+
+	// Set root for next iteration.
+	root := api.Select(skipFlag, gadget.Root, rootPostDeletion) // If skipFlag is set, we don't update the root.
+	return root
+}
+
+type DeletionProof struct {
+	DeletionIndices []frontend.Variable
+	PreRoot         frontend.Variable
+	IdComms         []frontend.Variable
+	MerkleProofs    [][]frontend.Variable
+
+	BatchSize int
+	Depth     int
+}
+
+func (gadget DeletionProof) DefineGadget(api abstractor.API) interface{} {
+	// Actual batch merkle proof verification.
+	root := gadget.PreRoot
+
+	// Individual deletions.
+	for i := 0; i < gadget.BatchSize; i += 1 {
+		// Set root for next iteration.
+		root = extractor.Call(api, DeletionRound{
+			Root: root,
+			Index: gadget.DeletionIndices[i],
+			Item: gadget.IdComms[i],
+			MerkleProofs: gadget.MerkleProofs[i],
+			Depth: gadget.Depth,
+		})
+	}
+
+	return root
+}
+
+type CheckBitOne struct {
```

Comment:
Any chance we could get rid of this (and `CheckBitZero`)? These gadgets feel very artificial and would look much better inlined (they reduce the readbility of the reducedness check, not improve it).